java/src/main/native/src/row_conversion.cu

```diff
@@ -304,20 +334,600 @@ copy_from_fixed_width_columns(const cudf::size_type start_row, const cudf::size_
   }
 }
 
+#if !defined(__CUDA_ARCH__) || __CUDA_ARCH__ >= 700
+
+/**
+ * @brief The GPU blocks work on one or more block_info structs of data.
+ *        This structure defined the workspace for the block.
+ *
+ */
+struct block_info {
+  int start_col;
+  int start_row;
+  int end_col;
+  int end_row;
+  int batch_number;
+
+  constexpr size_type get_shared_row_size(size_type const *const col_offsets,
+                                          size_type const *const col_sizes) const {
+    return util::round_up_unsafe(col_offsets[end_col] + col_sizes[end_col] - col_offsets[start_col],
+                                 JCUDF_ROW_ALIGNMENT);
+  }
+  constexpr size_type num_cols() const { return end_col - start_col + 1; }
+
+  constexpr size_type num_rows() const { return end_row - start_row + 1; }
+};
+
+/**
+ * @brief Returning rows is done in a byte cudf column. This is limited in size by
+ *        `size_type` and so output is broken into batches of rows that fit inside
+ *        this limit.
+ *
+ */
+struct row_batch {
+  size_type num_bytes;                   // number of bytes in this batch
+  size_type row_count;                   // number of rows in the batch
+  device_uvector<size_type> row_offsets; // offsets column of output cudf column
+};
+
+/**
+ * @brief Holds information about the batches of data to be processed
+ *
+ */
+struct batch_data {
+  device_uvector<size_type> batch_row_offsets; // offset column of returned cudf column
+  std::vector<size_type>
+      batch_row_boundaries;           // row numbers for different batches like 0, 10000, 20000
+  std::vector<row_batch> row_batches; // information about each batch such as byte count
+};
+
+/**
+ * @brief copy data from cudf columns into JCUDF format, which is row-based
+ *
+ * @param num_rows total number of rows in the table
+ * @param num_columns total number of columns in the table
+ * @param shmem_used_per_block shared memory amount each `block_info` is using
+ * @param block_infos span of `block_info` structs the define the work
+ * @param input_data pointer to raw table data
+ * @param col_sizes array of sizes for each element in a column - one per column
+ * @param col_offsets offset into input data row for each column's start
+ * @param row_offsets offset to a specific row in the output data
+ * @param output_data pointer to output data
+ *
+ */
+__global__ void copy_to_rows(const size_type num_rows, const size_type num_columns,
+                             const size_type shmem_used_per_block,
+                             device_span<const block_info> block_infos, const int8_t **input_data,
+                             const size_type *col_sizes, const size_type *col_offsets,
+                             const size_type *row_offsets, int8_t **output_data) {
```

Comment:
Since the size of each row is a constant in the all-fixed-width case, you should make `row_offsets` an iterator and just pass a transform iterator that does the math for the fixed-width case.

Reply:
I started to implement one of these iterators and it isn't as straight-forward as it would look. You don't just return `row_num * bytes_per_row` as you would expect, but instead you have to figure out what batch a row is in and subtract off the starting row. This turns into a `lower_bound`, a look up into global memory for the starting row of the batch, a subtraction, and then a multiplication. The alternative is a single global lookup. 😞  I'll think more on this, but as it sits this won't be an easy fix. The best I can come up with so far is to launch the kernel for each batch, which should be 1 or, in extreme cases, 2 launches. Then each launch could use that simple math.

Reply:
We talked about this and inverted the problem a bit. I was too stuck on the iterator simply taking a row number and magically knowing everything, but the kernel had the batch number available, so passing that in helped resolve this issue. This is updated. It did provide a marked speedup, but not as amazing as I was hoping.